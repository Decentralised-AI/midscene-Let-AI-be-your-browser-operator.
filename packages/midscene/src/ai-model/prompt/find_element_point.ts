import { call, callToGetJSONObject } from '@/ai-model/openai/index';
import type { AIUsageInfo, PlanningAIResponse, UIContext } from '@/types';
import { AIActionType } from '../common';

export const findElementPointPrompt = `
Based on the screenshot of the page, I give a text description and you give its corresponding center point location. The coordinate represents the center point [x, y] of the target element, which is a relative coordinate on the screenshot, scaled from 0 to 1. Please ensure the coordinate is at the exact center of the element.
`;
export async function findElementPoint<T>(
  userPrompt: string,
  opts: {
    screenshotBase64: string;
  },
) {
  const { screenshotBase64 } = opts;
  const data = await callToGetJSONObject(
    [
      { role: 'system', content: findElementPointPrompt },
      {
        role: 'user',
        content: [
          {
            //@ts-ignore
            type: 'image_url',
            image_url: {
              url: screenshotBase64,
            },
          },
          {
            type: 'text',
            text: `用户希望达成的目标：${userPrompt}`,
          },
        ],
      },
    ],
    AIActionType.INSPECT_ELEMENT,
  );
  return data as T;
}
