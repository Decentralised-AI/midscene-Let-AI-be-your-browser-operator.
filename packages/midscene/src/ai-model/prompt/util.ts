import assert from 'node:assert';
import { MATCH_BY_POSITION, getAIConfig } from '@/env';
import { imageInfoOfBase64 } from '@/image';
import type { BaseElement, ElementTreeNode, Size, UIContext } from '@/types';
import { PromptTemplate } from '@langchain/core/prompts';
import { NodeType } from '@midscene/shared/constants';
import { generateHashId } from '@midscene/shared/utils';
import type { ResponseFormatJSONSchema } from 'openai/resources';

const characteristic =
  'You are a versatile professional in software UI design and testing. Your outstanding contributions will impact the user experience of billions of users.';

const contextFormatIntro = `
The user will give you a screenshot and some of the texts on it. There may be some none-English characters (like Chinese) on it, indicating it's an non-English app. If some text is shown on screenshot but not introduced by the JSON description, use the information you see on screenshot.`;

export function systemPromptToLocateElement(
  queryPrompt: string,
  multi: boolean,
) {
  assert(queryPrompt, 'queryPrompt is required');
  return `
${characteristic}
${contextFormatIntro}

Based on the information you get, find ${
    multi ? 'one or more text elements' : 'ONE text element'
  } on the page.
Here is the description: ${queryPrompt}
  
Return in the following JSON format:
{
  "elements": [ // Leave it an empty array when no element is found
    { 
      "id": "id of the element, like 123", 
    },
    // more ...
  ], 
  errors?: [], // string[], error message if any
}
`;
}

export function systemPromptToExtract() {
  return `
You are a versatile professional in software UI design and testing. Your outstanding contributions will impact the user experience of billions of users.
The user will give you a screenshot and the contents of it. There may be some none-English characters (like Chinese) on it, indicating it's an non-English app.

You have the following skills:

skill name: extract_data_from_UI
related input: DATA_DEMAND
skill content: 
* User will give you some data requirements in DATA_DEMAND. Consider the UI context, follow the user's instructions, and provide comprehensive data accordingly.
* There may be some special commands in DATA_DEMAND, please pay extra attention
  - LOCATE_ONE_ELEMENT and LOCATE_ONE_OR_MORE_ELEMENTS: if you see a description that mentions the keyword LOCATE_ONE_ELEMENT
  - LOCATE_ONE_OR_MORE_ELEMENTS(e.g. follow LOCATE_ONE_ELEMENT : i want to find ...), it means user wants to locate a specific element meets the description. 

Return in this way: prefix + the id / comma-separated ids, for example: LOCATE_ONE_ELEMENT/1 , LOCATE_ONE_OR_MORE_ELEMENTS/1,2,3 . If not found, keep the prefix and leave the suffix empty, like LOCATE_ONE_ELEMENT/ .

Return in the following JSON format:
{
  language: "en", // "en" or "zh", the language of the page. Use the same language to describe section name, description, and similar fields.
  data: any, // the extracted data from extract_data_from_UI skill. Make sure both the value and scheme meet the DATA_DEMAND.
  errors: [], // string[], error message if any
}
`;
}

export const extractDataPrompt = new PromptTemplate({
  template: `
pageDescription: {pageDescription}

Use your extract_data_from_UI skill to find the following data, placing it in the \`data\` field
DATA_DEMAND start:
=====================================
{dataKeys}

{dataQuery}
=====================================
DATA_DEMAND ends.
  `,
  inputVariables: ['pageDescription', 'dataKeys', 'dataQuery'],
});

export const extractDataSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'extract_data',
    strict: true,
    schema: {
      type: 'object',
      properties: {
        language: {
          type: 'string',
          enum: ['en', 'zh'],
          description: 'The language of the page',
        },
        data: {
          type: 'object',
          description: 'The extracted data from extract_data_from_UI skill',
        },
        errors: {
          type: 'array',
          items: {
            type: 'string',
          },
          description: 'Error messages, if any',
        },
      },
      required: ['language', 'data', 'errors'],
      additionalProperties: false,
    },
  },
};

export function systemPromptToAssert() {
  return `
${characteristic}
User will give an assertion, and some information about the page. Based on the information you get, tell whether the assertion is truthy.

Return in the following JSON format:
{
  thought: string, // string, the thought of the assertion. Should in the same language as the assertion.
  pass: true, // true or false, whether the assertion is truthy
}
`;
}

export const assertSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'assert',
    strict: true,
    schema: {
      type: 'object',
      properties: {
        thought: {
          type: 'string',
          description: 'The thought process behind the assertion',
        },
        pass: {
          type: 'boolean',
          description: 'Whether the assertion passed or failed',
        },
      },
      required: ['thought', 'pass'],
      additionalProperties: false,
    },
  },
};

export function describeSize(size: Size) {
  return `${size.width} x ${size.height}`;
}

export function describeTextFormat() {
  // use `right` and `bottom` to help AI reduce the feasibility of performing computations
  return `
The following texts elements are formatted in the following way: 
id(string), left, top, right, bottom, content(may be truncated)`;
}

type PromptElementType = {
  id: BaseElement['id'];
  attributes: BaseElement['attributes'];
  rect: BaseElement['rect'];
  content: BaseElement['content'];
};

export function describeElement(
  elements: (Pick<BaseElement, 'rect' | 'content'> & { id: string })[],
) {
  const sliceLength = 80;
  return elements
    .map((item) =>
      [
        item.id,
        item.rect.left,
        item.rect.top,
        item.rect.left + item.rect.width,
        item.rect.top + item.rect.height,
        item.content.length > sliceLength
          ? `${item.content.slice(0, sliceLength)}...`
          : item.content,
      ].join(', '),
    )
    .join('\n');
}

export function truncateText(
  text: string | number | object | undefined,
  maxLength = 150,
) {
  if (typeof text === 'undefined') {
    return '';
  }

  if (typeof text === 'object') {
    text = JSON.stringify(text);
  }

  if (typeof text === 'number') {
    return text.toString();
  }

  if (typeof text === 'string' && text.length > maxLength) {
    return `${text.slice(0, maxLength)}...`;
  }

  if (typeof text === 'string') {
    return text.trim();
  }

  return '';
}

export function elementByPositionWithElementInfo(
  treeRoot: ElementTreeNode<BaseElement>,
  position: {
    x: number;
    y: number;
  },
) {
  assert(typeof position !== 'undefined', 'position is required for query');

  const matchingElements: BaseElement[] = [];

  function dfs(node: ElementTreeNode<BaseElement>) {
    if (node?.node) {
      const item = node.node;
      if (
        item.attributes.nodeType !== NodeType.CONTAINER &&
        item.rect.left <= position.x &&
        position.x <= item.rect.left + item.rect.width &&
        item.rect.top <= position.y &&
        position.y <= item.rect.top + item.rect.height
      ) {
        matchingElements.push(item);
      }
    }

    for (const child of node.children) {
      dfs(child);
    }
  }

  dfs(treeRoot);

  if (matchingElements.length === 0) {
    return undefined;
  }

  // Find the smallest element by area
  return matchingElements.reduce((smallest, current) => {
    const smallestArea = smallest.rect.width * smallest.rect.height;
    const currentArea = current.rect.width * current.rect.height;
    return currentArea < smallestArea ? current : smallest;
  });
}

export const samplePageDescription = `
The size of the page: 1280 x 720
Some of the elements are marked with a rectangle in the screenshot, some are not.

Description of all the elements in screenshot:
<div id="969f1637" markerId="1"> // The markerId indicated by the rectangle label in the screenshot
  <h4 id="b211ecb2" markerId="5" >
    The username is accepted
  </h4>
  ...many more
</div>
`;

export function trimAttributes(
  attributes: Record<string, any>,
  truncateTextLength?: number,
) {
  const tailorAttributes = Object.keys(attributes).reduce(
    (res, currentKey: string) => {
      const attributeVal = (attributes as any)[currentKey];
      if (
        currentKey === 'style' ||
        currentKey === 'src' ||
        currentKey === 'htmlTagName' ||
        currentKey === 'nodeType'
      ) {
        return res;
      }

      res[currentKey] = truncateText(attributeVal, truncateTextLength);
      return res;
    },
    {} as BaseElement['attributes'],
  );
  return tailorAttributes;
}

const nodeSizeThreshold = 4;
export async function descriptionOfTree<
  ElementType extends BaseElement = BaseElement,
>(
  tree: ElementTreeNode<ElementType>,
  truncateTextLength?: number,
  filterNonTextContent = false,
) {
  const attributesString = (kv: Record<string, any>) => {
    return Object.entries(kv)
      .map(
        ([key, value]) => `${key}="${truncateText(value, truncateTextLength)}"`,
      )
      .join(' ');
  };

  function buildContentTree(
    node: ElementTreeNode<ElementType>,
    indent = 0,
  ): string {
    let before = '';
    let contentWithIndent = '';
    let after = '';
    let emptyNode = true;
    const indentStr = '  '.repeat(indent);

    let children = '';
    for (let i = 0; i < (node.children || []).length; i++) {
      const childContent = buildContentTree(node.children[i], indent + 1);
      if (childContent) {
        children += `\n${childContent}`;
      }
    }

    if (
      node.node &&
      node.node.rect.width > nodeSizeThreshold &&
      node.node.rect.height > nodeSizeThreshold &&
      (!filterNonTextContent || (filterNonTextContent && node.node.content))
    ) {
      emptyNode = false;
      let nodeTypeString: string;
      if (node.node.attributes?.htmlTagName) {
        nodeTypeString = node.node.attributes.htmlTagName.replace(/[<>]/g, '');
      } else {
        nodeTypeString = node.node.attributes.nodeType
          .replace(/\sNode$/, '')
          .toLowerCase();
      }
      const markerId = (node.node as any).indexId;
      const markerIdString = markerId ? `markerId="${markerId}"` : '';
      const rectAttribute = node.node.rect
        ? {
            left: node.node.rect.left,
            top: node.node.rect.top,
            width: node.node.rect.width,
            height: node.node.rect.height,
          }
        : {};
      before = `<${nodeTypeString} id="${node.node.id}" ${markerIdString} ${attributesString(trimAttributes(node.node.attributes || {}, truncateTextLength))} ${attributesString(rectAttribute)}>`;
      const content = truncateText(node.node.content, truncateTextLength);
      contentWithIndent = content ? `\n${indentStr}  ${content}` : '';
      after = `</${nodeTypeString}>`;
    } else if (!filterNonTextContent) {
      if (!children.trim().startsWith('<>')) {
        before = '<>';
        contentWithIndent = '';
        after = '</>';
      }
    }

    if (emptyNode && !children.trim()) {
      return '';
    }

    const result = `${indentStr}${before}${contentWithIndent}${children}\n${indentStr}${after}`;
    if (result.trim()) {
      return result;
    }
    return '';
  }

  return buildContentTree(tree);
}

export async function describeUserPage<
  ElementType extends BaseElement = BaseElement,
>(
  context: Omit<UIContext<ElementType>, 'describer'>,
  opt?: {
    truncateTextLength?: number;
    filterNonTextContent?: boolean;
  },
) {
  const { screenshotBase64 } = context;
  let width: number;
  let height: number;

  if (context.size) {
    ({ width, height } = context.size);
  } else {
    const imgSize = await imageInfoOfBase64(screenshotBase64);
    ({ width, height } = imgSize);
  }

  const treeRoot = context.tree;
  // dfs tree, save the id and element info
  const idElementMap: Record<string, ElementType> = {};
  const flatElements: ElementType[] = [];
  function dfsTree(node: ElementTreeNode<ElementType>) {
    if (node?.node) {
      idElementMap[node.node.id] = node.node;
      flatElements.push(node.node);
    }
    for (let i = 0; i < (node.children || []).length; i++) {
      dfsTree(node.children[i]);
    }
  }
  dfsTree(treeRoot);

  const contentTree = await descriptionOfTree(
    treeRoot,
    opt?.truncateTextLength,
    opt?.filterNonTextContent,
  );

  // if match by position, don't need to provide the page description
  const pageJSONDescription = getAIConfig(MATCH_BY_POSITION)
    ? ''
    : `Some of the elements are marked with a rectangle in the screenshot, some are not. \n The page elements tree:\n${contentTree}`;
  const sizeDescription = describeSize({ width, height });

  return {
    description: `The size of the page: ${sizeDescription} \n ${pageJSONDescription}`,
    elementById(id: string) {
      assert(typeof id !== 'undefined', 'id is required for query');
      const item = idElementMap[`${id}`];
      return item;
    },
    elementByPosition(
      position: { x: number; y: number },
      size: { width: number; height: number },
    ) {
      console.log('elementByPosition', { position, size });
      return elementByPositionWithElementInfo(treeRoot, position);
    },
    insertElementByPosition(position: { x: number; y: number }) {
      const rect = {
        left: Math.max(position.x - 4, 0),
        top: Math.max(position.y - 4, 0),
        width: 8,
        height: 8,
      };
      const id = generateHashId(rect);
      const element = {
        id,
        attributes: { nodeType: NodeType.POSITION },
        rect,
        content: '',
        center: [position.x, position.y],
      } as ElementType;

      treeRoot.children.push({
        node: element,
        children: [],
      });
      flatElements.push(element);
      idElementMap[id] = element;
      return element;
    },
    size: { width, height },
  };
}
