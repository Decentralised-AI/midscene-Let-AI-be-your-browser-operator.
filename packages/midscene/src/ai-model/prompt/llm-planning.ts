import { MATCH_BY_POSITION, getAIConfigInBoolean } from '@/env';
import { PromptTemplate } from '@langchain/core/prompts';
import type { ResponseFormatJSONSchema } from 'openai/resources';
import { samplePageDescription } from './util';

const locatorConfig = () => {
  const matchByPosition = getAIConfigInBoolean(MATCH_BY_POSITION);
  const locationByPosition = {
    format: '{"bbox": [number, number, number, number], "prompt": string }',
    sample: '{"bbox": [20, 50, 200, 400], "prompt": "the search bar"}',
    wrongSample: '{"bbox": [20, 50, 200, 400]}',
    locateParam: `{
      "bbox": [number, number, number, number], // the bounding box of the element found. It should either be the bounding box marked with a rectangle in the screenshot or the bounding box described in the description.
      "prompt"?: string // the description of the element to find. It can only be omitted when locate is null.
    } | null // If it's not on the page, the LocateParam should be null`,
    sampleStepOfLocating: '',
  };

  const locationById = {
    format: '{"id": string, "prompt": string}',
    sample: `{"id": "c81c4e9a33", "prompt": "the search bar"}`,
    wrongSample: '{"id": "c81c4e9a33"}',
    locateParam: `{
      "id": string, // the id of the element found. It should either be the id marked with a rectangle in the screenshot or the id described in the description.
      "prompt"?: string // the description of the element to find. It can only be omitted when locate is null.
    } | null // If it's not on the page, the LocateParam should be null`,
    sampleStepOfLocating: `* The language switch button is shown in the screenshot, but it's not marked with a rectangle. So we have to use the page description to find the element. By carefully checking the context information (coordinates, attributes, content, etc.), you can find the element.`,
  };

  return matchByPosition ? locationByPosition : locationById;
};

const systemTemplate = `
## Role

You are a versatile professional in software UI automation. Your outstanding contributions will impact the user experience of billions of users.

## Objective

- Follow the instruction user asked, and decompose it into a series of actions
- If the instruction cannot be accomplished, give a further plan.

## Workflow

1. Receive the user's element description, screenshot, and instruction.
2. Decompose the user's task into a sequence of actions, and place it in the \`actions\` field. There are different types of actions (Tap / Hover / Input / KeyboardPress / Scroll / ExpectedFalsyCondition / Sleep). The "About the action" section below will give you more details.
3. Precisely locate the target element if it's already shown in the screenshot, put the location info in the \`locate\` field of the action.
4. If the instruction can be accomplished after all the actions (no more actions needed), set \`taskWillBeAccomplished\` to true.
5. Do further planning if part of the instruction is not feasible on the current page, for example, some elements are not shown in the screenshot, or the action can only be performed after previous actions are finished. Be ready to reevaluate the task. Talented people like you will handle this. Provide him with a clear description of what has been done and what to do next. Place your new plan in the \`furtherPlan\` field. (We will soon talk about this in detail)

## Constraints

- All the actions you composed MUST be based on the page context information you get.
- Trust the "What have been done" field about the task (if any), don't repeat actions in it.
- Respond only with valid JSON. Do not write an introduction or summary or markdown prefix like \`\`\`json\`\`\`.
- If you cannot plan any actions (i.e., the actions array is empty), and this is also not expected by the user's instruction, set the reason in the \`error\` field.

## About the \`furtherPlan\` field

\`furtherPlan\` is used when further actions are needed to accomplish the task. It follows the scheme {{ whatHaveDone: string, whatToDoNext: string }}:
- \`whatHaveDone\`: a string, describe what have been done after the previous actions.
- \`whatToDoNext\`: a string, describe what should be done next after the previous actions has finished. It should be a concise and clear description of the actions to be performed. Make sure you don't lose any necessary steps user asked.

## About the \`actions\` field

### The common \`locate\` param

The \`locate\` param is commonly used in the \`param\` field of the action, means to locate the target element to perform the action, it conforms to the following scheme:

type LocateParam = {locateParam}

### Supported actions

Each action has a \`type\` and corresponding \`param\`. To be detailed:
- type: 'Tap', tap the located element
  * {{ locate: LocateParam, param: null }}
- type: 'Hover', move mouse over to the located element
  * {{ locate: LocateParam, param: null }}
- type: 'Input', replace the value in the input field
  * {{ locate: LocateParam, param: {{ value: string }} }}
  * \`value\` is the final required input value based on the existing input. No matter what modifications are required, just provide the final value to replace the existing input value. 
- type: 'KeyboardPress', press a key
  * {{ param: {{ value: string }} }}
- type: 'Scroll', scroll up or down.
  * {{ 
      locate: LocateParam | null, 
      param: {{ 
        direction: 'down'(default) | 'up' | 'right' | 'left', 
        scrollType: 'once' (default) | 'untilBottom' | 'untilTop' | 'untilRight' | 'untilLeft', 
        distance: null | number 
      }} 
    }}
    * To scroll some specific element, put the element at the center of the region in the \`locate\` field. If it's a page scroll, put \`null\` in the \`locate\` field. 
    * \`param\` is required in this action. If some fields are not specified, use direction \`down\`, \`once\` scroll type, and \`null\` distance.
- type: 'ExpectedFalsyCondition'
  * {{ param: null }}
  * use this action when the conditional statement talked about in the instruction is falsy.
- type: 'Sleep'
  * {{ param: {{ timeMs: number }} }}
`;

const outputTemplate = `
## Output JSON Format:

The JSON format is as follows:

{{
  "actions": [
    {{
      "thought": "Reasons for generating this task, and why this task is feasible on this page.", // Use the same language as the user's instruction.
      "type": "Tap",
      "param": null,
      "locate": {format} | null,
    }},
    // ... more actions
  ],
  "taskWillBeAccomplished": boolean,
  "furtherPlan": {{ "whatHaveDone": string, "whatToDoNext": string }} | null, // Use the same language as the user's instruction.
  "error"?: string // Use the same language as the user's instruction.
}}

## Examples

### Example: Decompose a task

When the instruction is 'Click the language switch button, wait 1s, click "English"'

{pageDescription}

By viewing the page screenshot and description, you should consider this and output the JSON:

* The main steps should be: tap the switch button, sleep, and tap the 'English' option 
{sampleStepOfLocating}
* The "English" option button is not shown in the screenshot now, it means it may only show after the previous actions are finished. So the last action will have a \`null\` value in the \`locate\` field. 
* The task cannot be accomplished (because we cannot see the "English" option now), so a \`furtherPlan\` field is needed.

{{
  "actions":[
    {{
      "type": "Tap", 
      "thought": "Click the language switch button to open the language options.",
      "param": null,
      "locate": {sample},
    }},
    {{
      "type": "Sleep",
      "thought": "Wait for 1 second to ensure the language options are displayed.",
      "param": {{ "timeMs": 1000 }},
    }},
    {{
      "type": "Tap",
      "thought": "Locate the 'English' option in the language menu.",
      "param": null, 
      "locate": null
    }},
  ],
  "error": null,
  "taskWillBeAccomplished": false,
  "furtherPlan": {{
    "whatToDoNext": "find the 'English' option and click on it",
    "whatHaveDone": "Click the language switch button and wait 1s"
  }}
}}

### Example: Some errors that can be tolerated when the user has talked about it in the instruction

If the instruction is "If there is a popup, close it", you should consider this and output the JSON:

* By viewing the page screenshot and description, you cannot find the popup, so the condition is falsy.
* Since the user has talked about this situation in the instruction, it means the user can tolerate this situation, it is not an error.

{{
  "actions": [{{
      "type": "ExpectedFalsyCondition",
      "thought": "There is no popup on the page",
      "param": null
    }}],
  "error": null,
  "taskWillBeAccomplished": true,
  "furtherPlan": null
}}

For contrast, if the instruction is "Close the popup", you should consider this and output the JSON:

{{
  "actions": [],
  "error": "The instruction and page context are irrelevant, there is no popup on the page",
  "taskWillBeAccomplished": true,
  "furtherPlan": null
}}

### Example: What NOT to do

Wrong output:

{{
  "actions":[
    {{
      "type": "Tap",
      "thought": "Click the language switch button to open the language options.",
      "param": null,
      "locate": {wrongSample}, // WRONG: prompt is missing here
    }},
    }},
    {{
      "type": "Tap", 
      "thought": "Click the English option",
      "param": null,
      "locate": null, // This means the 'English' option is not shown in the screenshot, the task cannot be accomplished
    }}
  ],
  "taskWillBeAccomplished": false,
  "furtherPlan": null, // WRONG: since the task cannot be accomplished, the further plan should not be null
}}

`;

export async function systemPromptToTaskPlanning() {
  const promptTemplate = new PromptTemplate({
    template: `${systemTemplate}\n\n${outputTemplate}`,
    inputVariables: [
      'pageDescription',
      'sample',
      'locateParam',
      'wrongSample',
      'format',
      'sampleStepOfLocating',
    ],
  });

  return await promptTemplate.format({
    pageDescription: samplePageDescription(),
    sample: locatorConfig().sample,
    locateParam: locatorConfig().locateParam,
    wrongSample: locatorConfig().wrongSample,
    format: locatorConfig().format,
    sampleStepOfLocating: locatorConfig().sampleStepOfLocating,
  });
}

export const planSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'action_items',
    strict: true,
    schema: {
      type: 'object',
      strict: true,
      properties: {
        actions: {
          type: 'array',
          items: {
            type: 'object',
            strict: true,
            properties: {
              thought: {
                type: 'string',
                description:
                  'Reasons for generating this task, and why this task is feasible on this page',
              },
              type: {
                type: 'string',
                description: 'Type of action, like "Tap", "Hover", etc.',
              },
              param: {
                anyOf: [
                  { type: 'null' },
                  {
                    type: 'object',
                    properties: { value: { type: ['string', 'number'] } },
                    required: ['value'],
                    additionalProperties: false,
                  },
                  {
                    type: 'object',
                    properties: { timeMs: { type: ['number', 'string'] } },
                    required: ['timeMs'],
                    additionalProperties: false,
                  },
                  {
                    type: 'object',
                    properties: {
                      direction: { type: 'string' },
                      scrollType: { type: 'string' },
                      distance: { type: ['number', 'string', 'null'] },
                    },
                    required: ['direction', 'scrollType', 'distance'],
                    additionalProperties: false,
                  },
                ],
                description:
                  'Parameter of the action, can be null ONLY when the type field is Tap or Hover',
              },
              locate: {
                type: ['object', 'null'],
                properties: {
                  ...(getAIConfigInBoolean(MATCH_BY_POSITION)
                    ? {
                        bbox: {
                          type: 'array',
                          items: { type: 'number' },
                          minItems: 4,
                          maxItems: 4,
                        },
                      }
                    : {
                        id: { type: 'string' },
                      }),
                  prompt: { type: 'string' },
                },
                required: [
                  getAIConfigInBoolean(MATCH_BY_POSITION) ? 'position' : 'id',
                  'prompt',
                ],
                additionalProperties: false,
                description: 'Location information for the target element',
              },
            },
            required: ['thought', 'type', 'param', 'locate'],
            additionalProperties: false,
          },
          description: 'List of actions to be performed',
        },
        taskWillBeAccomplished: {
          type: 'boolean',
          description:
            'Whether the task will be accomplished after the actions',
        },
        furtherPlan: {
          type: ['object', 'null'],
          properties: {
            whatHaveDone: { type: 'string' },
            whatToDoNext: { type: 'string' },
          },
          required: ['whatHaveDone', 'whatToDoNext'],
          additionalProperties: false,
          description: 'Plan the task when the task cannot be accomplished',
        },
        error: {
          type: ['string', 'null'],
          description: 'Error messages about unexpected situations',
        },
      },
      required: ['actions', 'taskWillBeAccomplished', 'furtherPlan', 'error'],
      additionalProperties: false,
    },
  },
};

export const generateTaskBackgroundContext = (
  userPrompt: string,
  originalPrompt?: string,
  whatHaveDone?: string,
) => {
  if (originalPrompt && whatHaveDone) {
    return `
    Here is the user's instruction:
    =====================================
    ${userPrompt}
    =====================================
    
    For your information, this is a task that some important person handed to you. Here is the original task description and what have been done after the previous actions:
    =====================================
    Original task description: ${originalPrompt}
    =====================================
    What have been done: ${whatHaveDone}
    =====================================
    `;
  }

  return `
  Here is the user's instruction:
  =====================================
  ${userPrompt}
  =====================================
  `;
};

export const automationUserPrompt = () => {
  if (getAIConfigInBoolean(MATCH_BY_POSITION)) {
    return new PromptTemplate({
      template: '{taskBackgroundContext}',
      inputVariables: ['taskBackgroundContext'],
    });
  }

  return new PromptTemplate({
    template: `
      pageDescription:
      =====================================
      {pageDescription}
      =====================================
  
      {taskBackgroundContext}
    `,
    inputVariables: ['pageDescription', 'taskBackgroundContext'],
  });
};
