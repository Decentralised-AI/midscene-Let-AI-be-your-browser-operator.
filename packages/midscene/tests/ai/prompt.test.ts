import { systemPromptToTaskPlanning } from '@/ai-model/prompt/llm-planning';
import { descriptionOfTree } from '@/ai-model/prompt/util';
import { NodeType } from '@midscene/shared/constants';
import { describe, expect, it, test } from 'vitest';

describe('automation - computer', () => {
  it('should be able to generate prompt', async () => {
    const prompt = await systemPromptToTaskPlanning();
    // console.log(prompt);
    expect(prompt).toBeDefined();
  });
});

describe('utils', () => {
  const tree = {
    node: {
      attributes: {
        nodeType: NodeType.CONTAINER,
      },
      id: '1',
      indexId: 19,
      rect: {
        x: 0,
        y: 0,
        width: 100,
        height: 100,
        left: 0,
        top: 0,
      },
      center: [50, 50] as [number, number],
      content:
        'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
    },
    children: [
      {
        node: {
          attributes: {
            nodeType: NodeType.TEXT,
            ariaLabel:
              'image description, it could be a long text, very loooooooooooooooooooooooooooooooooooooooooong',
          },
          id: '2',
          indexId: 999,
          rect: {
            x: 0,
            y: 0,
            width: 100,
            height: 100,
            left: 0,
            top: 0,
          },
          center: [50, 50] as [number, number],
          content: 'world',
        },
        children: [],
      },
      {
        node: {
          attributes: {
            nodeType: NodeType.IMG,
            style: 'width: 100px; height: 100px;',
            src: 'https://example.com/image.jpg',
            htmlTagName: '<img>',
            ariaLabel: 'image description',
            storyContent:
              'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
          },
          id: '3',
          indexId: 20,
          rect: {
            x: 0,
            y: 0,
            width: 100,
            height: 100,
            left: 0,
            top: 0,
          },
          center: [50, 50] as [number, number],
          content: 'world 2345',
        },
        children: [
          {
            node: {
              attributes: {
                nodeType: NodeType.IMG,
                style: 'width: 100px; height: 100px;',
                src: 'https://example.com/image.jpg',
                htmlTagName: '<img>',
                ariaLabel: 'image description',
                storyContent:
                  'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
              },
              id: '3',
              indexId: 20,
              rect: {
                x: 0,
                y: 0,
                width: 100,
                height: 100,
                left: 0,
                top: 0,
              },
              center: [50, 50] as [number, number],
              content: '',
            },
            children: [],
          },
          {
            node: {
              attributes: {
                nodeType: NodeType.IMG,
                style: 'width: 100px; height: 100px;',
                src: 'https://example.com/image.jpg',
                htmlTagName: '<img>',
                ariaLabel: 'image description',
                storyContent:
                  'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
              },
              id: '3',
              indexId: 20,
              rect: {
                x: 0,
                y: 0,
                width: 3,
                height: 3,
                left: 0,
                top: 0,
              },
              center: [50, 50] as [number, number],
              content: '',
            },
            children: [],
          },
          {
            node: {
              attributes: {},
              id: '3',
              indexId: 20,
              rect: {
                x: 0,
                y: 0,
                width: 3,
                height: 3,
                left: 0,
                top: 0,
              },
              center: [50, 50] as [number, number],
              content: '',
            },
            children: [
              {
                node: {
                  attributes: {},
                  id: '3',
                  indexId: 20,
                  rect: {
                    x: 0,
                    y: 0,
                    width: 3,
                    height: 3,
                    left: 0,
                    top: 0,
                  },
                  center: [50, 50] as [number, number],
                  content: '',
                },
                children: [
                  {
                    node: {
                      attributes: {
                        nodeType: NodeType.IMG,
                        style: 'width: 100px; height: 100px;',
                        src: 'https://example.com/image.jpg',
                        htmlTagName: '<img>',
                        ariaLabel: 'image description',
                        storyContent:
                          'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
                      },
                      id: '3222',
                      indexId: 20,
                      rect: {
                        x: 0,
                        y: 0,
                        width: 100,
                        height: 100,
                        left: 0,
                        top: 0,
                      },
                      center: [50, 50] as [number, number],
                      content: 'world 2345',
                    },
                    children: [],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  };

  it('should be able to describe tree', async () => {
    const description = await descriptionOfTree(tree);
    console.log(description);
    expect(description).toMatchSnapshot();
  });

  it('should be able to describe tree, filterNonTextContent = true', async () => {
    const description = await descriptionOfTree(tree, 20, true);
    console.log(description);
    expect(description).toMatchSnapshot();
  });
});
