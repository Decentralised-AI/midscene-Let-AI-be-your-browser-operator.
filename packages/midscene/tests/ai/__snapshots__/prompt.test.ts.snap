// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > should be able to describe tree 1`] = `
"<container id="1" markerId="19" >
  Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man ...
  <text id="2" markerId="999" ariaLabel="image description, it could be a long text, very loooooooooooooooooooooooooooooooooooooooooong">
    world
  </text>
  <img id="3" markerId="20" ariaLabel="image description" storyContent="Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man ...">
    world 2345
    <img id="3" markerId="20" ariaLabel="image description" storyContent="Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man ...">
    </img>
    <>
    </>
  </img>
</container>"
`;

exports[`utils > should be able to describe tree, filterNonTextContent = true 1`] = `
"<container id="1" markerId="19" >
  Legend had it that t...
  <text id="2" markerId="999" ariaLabel="image description, i...">
    world
  </text>
  <img id="3" markerId="20" ariaLabel="image description" storyContent="Legend had it that t...">
    world 2345
  </img>
</container>"
`;
